Support requests.response.raw being a file-like object

IlyaSukhanov opened the pull request:
Previously httpie relied on requests.models.Response.raw being
urllib3.HTTPResponse.  The Requests documentation specifies that
[requests.models.Response.raw](https://docs.python-requests.org/en/master/api/#requests.Response.raw)
is a File-like object but allows for other types for internal use.

This change introduces graceful handling for scenarios when
requests.models.Response.raw is not urllib3.HTTPResponse. In such a scenario
httpie now falls back to extracting metadata from requests.models.Response
directly instead of direct access from protected protected members such as
response.raw._original_response. A side effect in this fallback procedure is
that we can no longer determine HTTP protocol version and report it as `??`.

This change is necessary to make it possible to implement TransportPlugins
without having to also needing to emulate internal behavior of urlib3 and
http.client.

This pull requests is an attempt to address https://github.com/httpie/httpie/issues/1093

IlyaSukhanov commented on httpie/models.py:
```diff
@@ -61,12 +61,27 @@ def headers(self):
         }[original.version]
 
-        status_line = f'HTTP/{version} {original.status} {original.reason}'
+        return version, original.status, original.reason, original.msg._headers
```
Status, reason and headers ought to be pulled from self._orig as is done in _headers_requests_response instead of accessing the protected member self._orig.raw._original_response. Its not clear why this was not done to begin with and for that reason I'm leaving the old logic in place.

Notable that same approach cannot be used for version. Perhaps it would be better to file a bug against requests and ask for HTTP version to be exposed without having to rely on protected members.

IlyaSukhanov replied:
If its reasonable to assume that
```
status = self._orig.raw._original_response.status_code
reason = self._orig.raw._original_response.reason
headers = self._orig.raw._original_response.msg._headers
```
Is effectively the same as.
```
status = self._orig.status_code
reason = self._orig.reason
headers = self._orig.headers.items()
```

Then its worthwhile to consider the slightly simpler implementation https://github.com/IlyaSukhanov/httpie/commit/c14deea1a68c979f7700be1217512ee39c06d742

BoboTiG replied:
I would be in favor of the slightly simpler implementation but I would like @jakubroztocil POV, if possible.

IlyaSukhanov replied:
Sounds good. Let me know if you'd prefer I open that as a separate pr or include it on this one.

BoboTiG replied:
Let's go with the simplified version :+1:

IlyaSukhanov replied:
Done

IlyaSukhanov commented on httpie/client.py:
```diff
@@ -104,10 +104,15 @@ def collect_messages(
+            if isinstance(response.raw, urllib3.HTTPResponse):
+                # noinspection PyProtectedMember
+                expired_cookies += get_expired_cookies(
+                    headers=response.raw._original_response.msg._headers
```
Its not entirely clear to me why `response.raw._original_response.msg._headers` is being utilized here instead of `response.headers`. I did run tests without making use of  `response.raw._original_response.msg._headers` and they passed. Logic is left in place to ensure compatibility for what ever case this was intended for, however would be happy to remove on recommendation from someone more experienced with the project.

IlyaSukhanov replied:
Upon further reflection and re-reading https://github.com/httpie/httpie/pull/929#discussion_r439017110 its probably because Requests represents headers in a dict instead of a tuple list. Headers with same name would be lost.

IlyaSukhanov replied:
Upon further experimentation it turns out its possible to not make use of protected _original for purposes of get_expired_cookies(). Here is an passing implementation https://github.com/IlyaSukhanov/httpie/commit/36daab668c8357d61dd27a954ab8aae556eca202. Can add this commit to this PR if desired.

BoboTiG replied:
The last iteration on your branch seems reasonable :+1: 
Could you merge it there please?

IlyaSukhanov replied:
Done

BoboTiG commented on httpie/utils.py:
```diff
@@ -96,16 +99,17 @@ def is_expired(expires: Optional[float]) -> bool:
     ]
 
     _max_age_to_expires(cookies=cookies, now=now)
+    "Set-Cookie: foo=bar; Expires=Mon, 21 Jun 2021 04:20:21 GMT; Max-Age=1; Path=/"
```
Is it a comment left for debugging or to expose the cookie syntax?

IlyaSukhanov replied:
Note to self that I forgot to delete.

BoboTiG reviewed (commented):
That seems promising :muscle: 

Let's wait on Jakub for one specific area of the patch.
In the meantime, could you stick with the project quotes guidelines (`"` -> `'`)?

IlyaSukhanov commented on httpie/models.py:
```diff
@@ -61,12 +61,25 @@ def headers(self):
+
+    def _headers_requests_response(self):
+        """ Extract header information from Requests.models.Response """
+        version = "??"
```
Seems like a reasonable default. Done!

IlyaSukhanov commented on httpie/models.py:
```diff
@@ -61,12 +61,25 @@ def headers(self):
-        # `_headers` is a 2-tuple
         headers.extend(
-            f'{header[0]}: {header[1]}' for header in original.msg._headers)
+            ": ".join(header) for header in http_headers)
```
quotes fixed

IlyaSukhanov commented:
Thank for for review and feedback.

Latest changes:
* fixed quotes for consistency
* in unknown case http protocol version is set to 1.1
* removed stray debug comment

BoboTiG commented on Makefile:
```diff
@@ -77,14 +77,20 @@ venv:
 test:
 	@echo $(H1)Running tests$(HEADER_EXTRA)$(H1END)
-	$(VENV_BIN)/python -m pytest $(COV) ./httpie $(COV) ./tests --doctest-modules --verbose ./httpie ./tests
+	$(VENV_BIN)/python -m pytest $(COV) ./httpie $(COV) ./tests --doctest-modules --verbose ./httpie ./tests $(MISSING)
```
Could you revert those changes and propose them in a specific PR if you think it is interesting to keep :pray: ?

IlyaSukhanov replied:
certainly

IlyaSukhanov replied:
Done

IlyaSukhanov commented:
* Make file alterations have been reverted.
* Simplified envelope extraction added.

BoboTiG commented on httpie/models.py:
```diff
@@ -52,21 +52,23 @@ def iter_lines(self, chunk_size):
+            raw_version = self._orig.raw._original_response.version
+        except AttributeError:
+            # Assume HTTP/1.1
+            raw_version = '1.1'
```
```suggestion
            raw_version = 11
```

jakubroztocil commented on tests/test_sessions.py:
```diff
@@ -368,11 +363,10 @@ def test_get_expired_cookies_using_max_age(self):
-                    ('Set-Cookie', 'pea=pod; Path=/ab; Expires=Thu, 01-Jan-1970 00:00:00 GMT; HttpOnly'),
-                    ('Connection', 'keep-alive')
-                ],
+                (
+                    'hello=world; Path=/; Expires=Thu, 01-Jan-1970 00:00:00 GMT; HttpOnly, '
+                    'pea=pod; Path=/ab; Expires=Thu, 01-Jan-1970 00:00:00 GMT; HttpOnly'
+                ),
```
These should now be a `list`s.

IlyaSukhanov replied:
Cookies were stored as a list in `response.raw._original_response.msg._headers` however with this change cookies are extracted from `response.headers.get('Set-Cookie', '')` which is a comma separated string.

get_expired_cookies has been changed from accepting list-of-tuple headers to string of comma separated strings.

I don't see of a way to access more-or-less raw cookie data from requests in any way but parsing them out of the comma separated string without resorting to using protected members. Cookie jar has more formatted cookies but they are stored in a lossy way.

With all this in mind, to keep tests as lists the cookie splitting logic can be moved to outside of `get_expired_cookies()`, but I'm not sure if this is what you had in mind by your comment.

Thanks
--IAS

BoboTiG replied:
@IlyaSukhanov Is it an option to add a test case to ensure `response.headers.get('Set-Cookie', '')` returns a string with concatenated cookies? It will help us prevent potential regressions for upcoming `requests` versions and to be more comfortable with those changes too :)

After that, we will be good to merge.

IlyaSukhanov replied:
Added test to ensure that multiple cookies get extracted correctly.

IlyaSukhanov replied:
Looks like cicd checks is failing. Is there a tried approach for starting a local http server to test against?

The way it's erroring out it's not clear what the issue is. Few of suspicions I have:

1. Multiprocess does not start for some reason
2. Multiprocess does not start fast enough (single threaded machine?)
3. Flask is unable to open port for some reason (seems unlikely as no error to that effect is presented)


Without knowing what the error is or much experience with GitHub check infrastructure. My only idea is to add a retry functionality in test, to call httpie multiple time until hopefully the background server becomes responsive.

jakubroztocil replied:
I was referring to the incorrect usage of tuples for what by nature is a list. 

```python
(  # <= tuple because fixed-length collection where each item is a different thing

    ( # <= list because it’s a simple list of things
        'hello=world; Path=/; Expires=Thu, 01-Jan-1970 00:00:00 GMT; HttpOnly, '
        'pea=pod; Path=/ab; Expires=Thu, 01-Jan-1970 00:00:00 GMT; HttpOnly'
    ),
    None,
    [  # <= list because it’s a simple list of things
        {'name': 'hello', 'path': '/'},
        {'name': 'pea', 'path': '/ab'}
    ]
),
```

IlyaSukhanov replied:
I think I see what you mean. But the cookie values are not tuples its a string that is line wrapped, and thus grouped using parenthesis.

Tuple:
```
(
    'foo',
    'bar'
)
```

String:
```
(
    'foo'
    'bar'
)
```

Note how in this example and in the pr the lines are not separated by comma which is required in tuple definition.

I hope this clears up any confusion.

If project has different style requirement for line wrapping long strings, please let me know and I'll make this change conform to that.

jakubroztocil replied:
@IlyaSukhanov right, I see it now! Yeah, this style of string-wrapping is the preferred one, but in this context, next to a bunch of tuples, it’s easy to misread it (🙋‍♂️), so I’d vote for using a different style here.

jakubroztocil commented on CHANGELOG.rst:
```diff
@@ -13,6 +13,7 @@ This project adheres to `Semantic Versioning <https://semver.org/>`_.
   an alternative to ``stdin``. (`#534`_)
 * Fixed ``--continue --download`` with a single byte to be downloaded left. (`#1032`_)
 * Fixed ``--verbose`` HTTP 307 redirects with streamed request body. (`#1088`_)
+* Fixed support of file-like object responses. (`#1094`_)
```
```suggestion
* Add internal support for file-like object responses to improve adapter plugin support. (`#1094`_)
```

BoboTiG reviewed (commented):
Is there a way to craft the `Response` without using multiprocessing and `flask`?

IlyaSukhanov commented:
If the objective is to test internal of requests to ensure that it returns cookies in right format (and does not regress in future), the viable options I can think of are:

1. Call a local http server and have it respond with desired payload. This is what e57adec does. e57adec makes use of flask and multiprocessing to start the local http server. There are other options besides flask (`http.server.HTTPServer`) for this task, but it would still need to rely on multiprocess to ensure server and request can run in parallel.. and it would possibly be sort of fragile unless of course there are ready made solutions made for this that are thoroughly tested.
2. Mock downstream of `requests`; either `urllib3` or `socket` and pass up a fake payload through requests ensuring request handles it correctly. I suspect this is not a trivial implementation but could be made easier via something like https://github.com/mindflayer/python-mocket

Its also possible to mock `requests.Response` directly but it would not protect from internal implementation changes in requests down the road.

BoboTiG commented:
Maybe something "lighter" like done in https://github.com/getsentry/responses/blob/bdf4c3b1273025338c6a475a556d741552ffa819/responses/test_responses.py#L845 could be interesting?

I would love to see something using "raw" objects (from `urllib3`/`requests`/...) to simply tell the response there is several cookies. And then, check they all are processed as expected.

I do not have a strong opinion as soon as tests are green et possibly not too random.

BoboTiG commented on tests/test_cookie.py:
```diff
@@ -0,0 +1,37 @@
+
+def test_cookie_parser():
+    port = available_port()
+    server = Process(target=app.run, kwargs={'port': port})
```
```suggestion
    server = Process(target=app.run, kwargs={'port': 8888})
```

BoboTiG commented on tests/test_cookie.py:
```diff
@@ -0,0 +1,37 @@
+
+
+def test_cookie_parser():
+    port = available_port()
+    server = Process(target=app.run, kwargs={'port': 8888})
```
```suggestion
    port = 8080
    server = Process(target=app.run, kwargs={'port': port})
```

BoboTiG commented:
@IlyaSukhanov I'll handle the cookies test, it is a bonus and you've done so much already :+1:

BoboTiG commented on setup.py:
```diff
@@ -23,6 +23,7 @@
     'pytest-cov',
     'twine',
     'wheel',
+    'flask',
```
```suggestion
```

BoboTiG commented on tests/test_cookie.py:
```diff
@@ -0,0 +1,47 @@
+                """Handle GET requests."""
+                # Craft multiple cookies
+                cookie = SimpleCookie()
+                cookie['hello'] = 'world'
+                cookie['hello']['path'] = '/'
+                cookie['oatmeal_raisin'] = 'is the best'
+                cookie['oatmeal_raisin']['path'] = '/'
```
```suggestion
                cookie['hello'] = 'world'
                cookie['hello']['path'] = self.path
                cookie['oatmeal_raisin'] = 'is the best'
                cookie['oatmeal_raisin']['path'] = self.path
```

BoboTiG commented:
OK it should be good now. Do you mind having a final look at the test @IlyaSukhanov?

BoboTiG commented on httpie/models.py:
```diff
@@ -1,6 +1,8 @@
 from typing import Iterable, Optional
 from urllib.parse import urlsplit
 
+from httpie.utils import split_cookies
```
```suggestion
from .utils import split_cookies
```

BoboTiG commented on httpie/utils.py:
```diff
@@ -85,8 +88,21 @@ def get_content_type(filename):
 
+def split_cookies(cookies):
+    """
+    When Requests stores cookies in ``response.headers['Set-Cookie']``
+    it concatenates all of them through ``, ``
```
```suggestion
    When ``requests`` stores cookies in ``response.headers['Set-Cookie']``
    it concatenates all of them through ``, ``.
```

IlyaSukhanov commented:
Elegant!

An observation, and not a problem: The thread is not actively terminated. I'm not sure if pytest kills threads of between each test. Even if it doesn't I cannot think of a scenario where current implementation would become problematic, if more tests of this sort were added each test server would be isolated to its own thread/port. And obviously pytest finishes execution so it must be terminating the thread at least at the completion of all tests.

:+1: :fireworks:

BoboTiG commented:
Yes, about the zombie thread, it is not an issue right now. And I failed to find a multi-platform solution. It could be improved if we ever need more tests like that.

BoboTiG commented:
That's awesome, thanks a lot for your work and patience @IlyaSukhanov :pie: :champagne:
